sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala

```diff
@@ -304,8 +468,173 @@ class RapidsBufferCatalog extends AutoCloseable with Arm {
     bufferMap.compute(buffer.id, updater)
   }
 
-  /** Remove a buffer ID from the catalog at the specified storage tier. */
-  def removeBufferTier(id: RapidsBufferId, tier: StorageTier): Unit = {
+  /**
+   * Free memory in `store` by spilling buffers to the spill store synchronously.
+   * @param store store to spill from
+   * @param targetTotalSize maximum total size of this store after spilling completes
+   * @param stream CUDA stream to use or null for default stream
+   * @return optionally number of bytes that were spilled, or None if this called
+   *         made no attempt to spill due to a detected spill race
+   */
+  def synchronousSpill(
+      store: RapidsBufferStore,
+      targetTotalSize: Long,
+      stream: Cuda.Stream = Cuda.DEFAULT_STREAM): Option[Long] = {
+    val spillStore = store.spillStore
+    if (spillStore == null) {
+      throw new OutOfMemoryError("Requested to spill without a spill store")
+    }
+    require(targetTotalSize >= 0, s"Negative spill target size: $targetTotalSize")
+    logWarning(s"Targeting a ${store.name} size of $targetTotalSize. " +
+      s"Current total ${store.currentSize}. " +
+      s"Current spillable ${store.currentSpillableSize}")
+
+    // we try to spill in this thread. If another thread is also spilling, we let that
+    // thread win and we return letting RMM retry the alloc
+    var rmmShouldRetryAlloc = false
+
+    // total amount spilled in this invocation
+    var totalSpilled: Long = 0
+
+    if (store.currentSpillableSize > targetTotalSize) {
```

Comment:
Not necessarily to fix for this PR, but one issue with this old logic is that it will not free _anything_ if targetTotalSize is larger than the current spillable size.  That seems pretty bad.  For example, freeing just 256MB of memory might allow a 1GB allocation to succeed.  Honestly not sure why we would care what the current amount of spillable memory is, we should just blindly ask the store to spill until we can allocate, and the store either will spill or it won't.  We shouldn't give up until the spill store is exhausted of spillable memory or we succeed in the allocation, whichever comes first.

Reply:
Filed this to track https://github.com/NVIDIA/spark-rapids/issues/7706